File: 2293/2293.js
var fs = require("fs");
const filepath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const splitStr = process.platform === "linux" ? "\n" : "\r\n";
var input = fs.readFileSync(filepath).toString().split(splitStr);
const [N, M] = input[0].split(" ").map(Number);
const coins = input.slice(1, 1 + N).map(Number);

function solution(N, M, coins) {
  let dp = Array(10005).fill(0);
  dp[0] = 1;

  coins.forEach((n) => {
    for (let i = 0; i <= 10000 - n; i++) {
      if (dp[i]) {
        dp[i + n] += dp[i];
      }
    }
  });
  return dp[N];
}

console.log(solution(N, M, coins));

File: 2293/2293.py
import sys
sys.stdin = open('input.txt')
input = sys.stdin.readline
n,k = map(int,input().split(' '))
coins = [int(input()) for _ in range(n)]


def solution(n,k,coins):
    dp = [0 for _ in range(10005)]
    dp[0] = 1
    for coin in coins:
        for n in range(0,10000-coin+1):
            if dp[n] : dp[n+coin] += dp[n]

    return dp[k]


print(solution(n,k,coins))


